### PR TITLE
mem: A unified memory controller for dram cache + nvm

### PR DESCRIPTION
**Pull request by aakahlow**
-  Many comments needs to be cleaned
-  There are still things TODO (mostly there should be some comments explaining what needs to be done)

### CODE REVIEW

File: configs/nvm/sweep_hybrid.py [hunk 1]
```diff
@@ -78,9 +78,9 @@
 parser.add_option("--nvm-perc", type="int", default=100,
                   help = "Percentage of NVM commands")
 
-parser.add_option("--mode", type="choice", default="HYBRID",
-                  choices=hybrid_generators.keys(),
-                  help = "Hybrid: Random DRAM + NVM traffic")
+#parser.add_option("--mode", type="choice", default="HYBRID",
```
Comment (powerjg) [inline configs/nvm/sweep_hybrid.py:81]: Why this change?

File: src/mem/MemCtrl.py [hunk 2]
```diff
@@ -58,6 +58,8 @@ class MemCtrl(QoSMemCtrl):
     # bus in front of the controller for multiple ports
     port = ResponsePort("This port responds to memory requests")
 
+    dram_cache_size = Param.Int('1024', "DRAM cache size")
```
Comment (powerjg) [inline src/mem/MemCtrl.py:61]: 1024 what? 

This param should probably be a "MemorySize"

File: src/mem/MemCtrl.py [hunk 3]
```diff
@@ -58,6 +58,8 @@ class MemCtrl(QoSMemCtrl):
...
+    write_allocate_policy = Param.Int('false', "DRAM cache write allocate policy, can be 'allocate on write' or 'no allocate on write'")
```
Comment (powerjg) [inline src/mem/MemCtrl.py:62]: Think this should be a bool.

Also, line length :) How were you able to commit?!?!?!?!?!?!

File: src/mem/MemCtrl.py [hunk 4]
```diff
@@ -89,4 +91,6 @@ class MemCtrl(QoSMemCtrl):
     static_frontend_latency = Param.Latency("10ns", "Static frontend latency")
     static_backend_latency = Param.Latency("10ns", "Static backend latency")
 
+    # latency for dram cache tag check
+    static_tagcheck_latency = Param.Latency("2ns", "Static tagcheck latency")
```
Comment (powerjg) [inline src/mem/MemCtrl.py:95]: The help string should be more than just repeating the name. (for instance, you could have put the comment on the line before as the help string)

File: src/mem/mem_ctrl.hh [hunk 5]
```diff
@@ -81,7 +80,8 @@ class BurstHelper
 
     BurstHelper(unsigned int _burstCount)
         : burstCount(_burstCount), burstsServiced(0)
-    { }
+    {
```
Comment (powerjg) [inline src/mem/mem_ctrl.hh:83]: unneeded change. I won't mark all of these, but please look at the diff on github or in VS code and see what's not important and revert those parts.

File: src/mem/mem_ctrl.hh [hunk 6]
```diff
@@ -106,8 +105,12 @@ class MemPacket
 
     const bool read;
 
+    //every write will initiate a read first
```
Comment (powerjg) [inline src/mem/mem_ctrl.hh:108]: Needs more comment here, at least if this is going to be the permanent way this is implemented.

Also, should probably be const

File: src/mem/mem_ctrl.hh [hunk 7]
```diff
@@ -106,8 +105,12 @@ class MemPacket
...
+    bool read_before_write;
+
     /** Does this packet access DRAM?*/
-    const bool dram;
+    //const bool dram;
```
Comment (powerjg) [inline src/mem/mem_ctrl.hh:112]: Why no longer const?

I think probably because you're re-using the same packet throughout the lifetime of the request. I am not convinced this is the best way to do it. I would have expected you to create a new DRAMPacket for each "step"

File: src/mem/mem_ctrl.hh [hunk 8]
```diff
@@ -139,7 +142,7 @@ class MemPacket
      * A pointer to the BurstHelper if this MemPacket is a split packet
      * If not a split packet (common case), this is set to NULL
      */
-    BurstHelper* burstHelper;
+    BurstHelper *burstHelper;
```
Comment (powerjg) [inline src/mem/mem_ctrl.hh:145]: Ok, this is actually the last time I'm going to comment on a non-functional change :)

File: src/mem/mem_ctrl.hh [hunk 9]
```diff
@@ -269,6 +276,40 @@ class MemCtrl : public QoS::MemCtrl
...
+    {
+    public:
+      Addr tag = -1;
+      Addr index = -1;
+      // xxxxxxdv (dirty and valid bits)
+      uint8_t meta_bits = 0;
+    };
+    // constant to indicate that the cache line is
+    // dirty
+    const uint8_t DIRTY_LINE = 1 << 0;
```
Comment (powerjg) [inline src/mem/mem_ctrl.hh:292]: I suggest not being so fancy and just adding more boolean members.

File: src/mem/mem_ctrl.hh [hunk 10]
```diff
@@ -269,6 +276,40 @@ class MemCtrl : public QoS::MemCtrl
      */
     MemoryPort port;
 
+    // Add a tag store map
+    // TODO: look at what the intialization
+    // values should be
+    class tag_entry
+    {
+    public:
+      Addr tag = -1;
```
Comment (powerjg) [inline src/mem/mem_ctrl.hh:285]: Addr of -1 doesn't make sense (it's an unsigned). I suggest using `0` and adding a `valid` field

File: src/mem/mem_ctrl.hh [hunk 11]
```diff
@@ -269,6 +276,40 @@ class MemCtrl : public QoS::MemCtrl
...
+    // dirty
+    const uint8_t DIRTY_LINE = 1 << 0;
+
+    // constant to indicate that the cache line is
+    // valid
+    const uint8_t VALID_LINE = 1 << 1;
+
+    // DC refers to Dram Cache
+    //uint64_t dramCacheSize;
+    uint64_t num_entries;
```
Comment (powerjg) [inline src/mem/mem_ctrl.hh:300]: camelCase (this really confused me in the constructor)

File: src/mem/mem_ctrl.hh [hunk 12]
```diff
@@ -269,6 +276,40 @@ class MemCtrl : public QoS::MemCtrl
      */
     MemoryPort port;
 
+    // Add a tag store map
+    // TODO: look at what the intialization
+    // values should be
+    class tag_entry
```
Comment (powerjg) [inline src/mem/mem_ctrl.hh:282]: CamelCase

File: src/mem/mem_ctrl.hh [hunk 13]
```diff
@@ -280,6 +321,21 @@ class MemCtrl : public QoS::MemCtrl
...
+    bool retryNVMWrReq;
+    bool retryDRAMFillReq;
+
+    /**
+     * Current sizes of nvmReadQueue, nvmWriteQueue, dramFillQueue
+     *
+     */
+    uint64_t nvmReadQueueSize;
+    uint64_t nvmWriteQueueSize;
+    uint64_t dramFillQueueSize;
```
Comment (powerjg) [inline src/mem/mem_ctrl.hh:338]: I suggest making these the *max* size and just doing queue.size() for the current size. (And make them const)

File: src/mem/mem_ctrl.hh [hunk 14]
```diff
@@ -308,6 +364,30 @@ class MemCtrl : public QoS::MemCtrl
      */
     bool writeQueueFull(unsigned int pkt_count) const;
 
+    /**
+     * Check if the nvm read queue has room for more entries
+     *
+     * @param pkt_count The number of entries needed in the nvm read queue
+     * @return true if nvm read queue is full, false otherwise
```
Comment (powerjg) [inline src/mem/mem_ctrl.hh:371]: Great documentation! Good job! :100:

File: src/mem/mem_ctrl.hh [hunk 15]
```diff
@@ -504,6 +619,12 @@ class MemCtrl : public QoS::MemCtrl
      */
     const Tick backendLatency;
 
+    /**
+     * This is the latency parameter to check
+     * DRAM cache tags.
+     */
+    const Tick tagcheckLatency;
```
Comment (powerjg) [inline src/mem/mem_ctrl.hh:626]: I'd use `tagCheckLatency` "tag check" is two words

File: src/mem/mem_interface.cc [hunk 16]
```diff
@@ -646,7 +646,7 @@ DRAMInterface::doBurstAccess(MemPacket* mem_pkt, Tick next_burst_at,
     }
 
     // DRAMPower trace command to be written
-    std::string mem_cmd = mem_pkt->isRead() ? "RD" : "WR";
+    std::string mem_cmd = (mem_pkt->isRead() || (!mem_pkt->isRead() && mem_pkt->isReadBeforeWrite())) ? "RD" : "WR";
```
Comment (powerjg) [inline src/mem/mem_interface.cc:649]: line length. Again, how were you allowed to commit?!

File: src/mem/qos/mem_ctrl.hh [hunk 17]
```diff
@@ -60,7 +60,7 @@ class MemCtrl : public ClockedObject
 {
   public:
     /** Bus Direction */
-    enum BusState { READ, WRITE };
+    enum BusState { READ, WRITE, ReadBeforeWrite };
```
Comment (powerjg) [inline src/mem/qos/mem_ctrl.hh:63]: Hmm... this clearly doesn't match the style.

Also "read before write" isn't a *bus state*. It's something else.

**Review by powerjg — COMMENTED** [review]
I haven't gone through the main logic, yet. Everything else looks pretty good beyond some style things.

File: configs/nvm/sweep_hybrid.py (hunk 1 above)
Comment (mbabaie) [reply configs/nvm/sweep_hybrid.py:81]: This file can be ignored. I was just reading it and trying to run it, but it was failing. So, I had to made this change to run.

File: src/mem/MemCtrl.py (hunk 2 above)
Comment (mbabaie) [reply src/mem/MemCtrl.py:61]: This is in Bytes.
Should it be '1024 B' or '1024'?

File: src/mem/mem_ctrl.hh [hunk 18]
```diff
@@ -269,6 +283,40 @@ class MemCtrl : public QoS::MemCtrl
...
+    // Add a tag store map
+    // TODO: look at what the intialization
+    // values should be
+    class tagEntry
+    {
+    public:
+      Addr tag = 0;
+      Addr index = 0;
+      // xxxxxxdv (dirty and valid bits)
+      uint8_t metadata = 0;
```
Comment (powerjg) [inline src/mem/mem_ctrl.hh:295]: Style
```suggestion
      public:
        Addr tag = 0;
        Addr index = 0;
        // xxxxxxdv (dirty and valid bits)
        uint8_t metadata = 0;
```

File: src/mem/mem_ctrl.hh [hunk 19]
```diff
@@ -269,6 +283,40 @@ class MemCtrl : public QoS::MemCtrl
...
+    // dirty
+    const uint8_t DIRTY_LINE = 1 << 0;
+
+    // constant to indicate that the cache line is
+    // valid
+    const uint8_t VALID_LINE = 1 << 1;
+
+    // DC refers to Dram Cache
+    //uint64_t dramCacheSize;
+    uint64_t numEntries;
```
Comment (powerjg) [inline src/mem/mem_ctrl.hh:307]: Number of blocks/entries in the DRAM cache

File: src/mem/mem_ctrl.hh [hunk 20]
```diff
@@ -269,6 +283,40 @@ class MemCtrl : public QoS::MemCtrl
      */
     MemoryPort port;
 
+    // Add a tag store map
+    // TODO: look at what the intialization
+    // values should be
+    class tagEntry
```
Comment (powerjg) [inline src/mem/mem_ctrl.hh:289]:
```suggestion
    class TagEntry
```

File: src/mem/mem_ctrl.hh [hunk 21]
```diff
@@ -525,6 +653,10 @@ class MemCtrl : public QoS::MemCtrl
      */
     Tick nextReqTime;
 
+    uint64_t dramCacheSize;
```
Comment (powerjg) [inline src/mem/mem_ctrl.hh:656]: I'd put this with `numEntries`

File: src/mem/mem_ctrl.cc [hunk 22]
```diff
@@ -89,6 +93,13 @@ MemCtrl::MemCtrl(const MemCtrlParams &p) :
         fatal("Write buffer low threshold %d must be smaller than the "
               "high threshold %d\n", p.write_low_thresh_perc,
               p.write_high_thresh_perc);
+
+    numEntries = ceilLog2(dramCacheSize/64);
```
Comment (powerjg) [inline src/mem/mem_ctrl.cc:97]: Move to initializer list

File: src/mem/mem_ctrl.cc [hunk 23]
```diff
@@ -89,6 +93,13 @@ MemCtrl::MemCtrl(const MemCtrlParams &p) :
...
+
+    for (int i = 0; i < numEntries; i++) {
+        tagEntry entry;
+        tagStoreDC.emplace_back(entry);
+    }
```
Comment (powerjg) [inline src/mem/mem_ctrl.cc:102]:
```suggestion
    tagStoreDC.resize(numEntries);
```

File: src/mem/mem_ctrl.cc [hunk 24]
```diff
@@ -187,24 +198,120 @@ MemCtrl::writeQueueFull(unsigned int neededEntries) const
...
+    return  wrsize_new > writeBufferSize;
+}*/
+
+bool
+MemCtrl::nvmWriteQueueFull(unsigned int neededEntries) const
+{
+    auto size = (nvmWriteQueueSize + neededEntries);
+    // random size to compare with for now
+    //return  size_new > 64;
+    return  size > 64;
```
Comment (powerjg) [inline src/mem/mem_ctrl.cc:229]: 64 should be a maxNvmWriteQueueSize parameter

File: src/mem/mem_ctrl.cc [hunk 25]
```diff
@@ -187,24 +198,120 @@ MemCtrl::writeQueueFull(unsigned int neededEntries) const
...
+            writeBufferSize, totalWriteQueueSize, neededEntries);
+
+    auto wrsize_new = (totalWriteQueueSize + neededEntries);
+    return  wrsize_new > writeBufferSize;
+}*/
+
+bool
+MemCtrl::nvmWriteQueueFull(unsigned int neededEntries) const
+{
+    auto size = (nvmWriteQueueSize + neededEntries);
```
Comment (powerjg) [inline src/mem/mem_ctrl.cc:226]:
```suggestion
    int size = (nvmWriteQueue.size() + neededEntries);
```

File: src/mem/mem_ctrl.cc [hunk 26]
```diff
@@ -187,24 +198,120 @@ MemCtrl::writeQueueFull(unsigned int neededEntries) const
...
+}
+
+
+bool
+MemCtrl::nvmReadQueueFull(unsigned int neededEntries) const
+{
+    auto size = (nvmReadQueueSize + neededEntries);
+    // random size to compare with for now
+    //return  size_new > 64;
+    return  size > 64;
```
Comment (powerjg) [inline src/mem/mem_ctrl.cc:239]: Ditto

File: src/mem/mem_ctrl.cc [hunk 27]
```diff
@@ -187,24 +198,120 @@ MemCtrl::writeQueueFull(unsigned int neededEntries) const
...
+}
+
+
+bool
+MemCtrl::dramFillQueueFull(unsigned int neededEntries) const
+{
+    auto size = (dramFillQueueSize + neededEntries);
+    // random size to compare with for now
+    //return  size_new > 64;
+    return  size > 64;
```
Comment (powerjg) [inline src/mem/mem_ctrl.cc:249]: Ditto

File: src/mem/mem_ctrl.cc [hunk 28]
```diff
@@ -187,24 +198,120 @@ MemCtrl::writeQueueFull(unsigned int neededEntries) const
...
+}
+
+void
+MemCtrl::addToNVMReadQueue(MemPacket* mem_pkt)
+{
+
+    // return true if this request can succeed
+    // false otherwise
+
+    // COMMENT: Do we need to snoop the write queue?
```
Comment (powerjg) [inline src/mem/mem_ctrl.cc:259]: I think so...

File: src/mem/mem_ctrl.cc [hunk 29]
```diff
@@ -187,24 +198,120 @@ MemCtrl::writeQueueFull(unsigned int neededEntries) const
...
+
+    // COMMENT: Can there be another read request
+    // to the same address in this queue
+    // If yes, then they should be merged
+
+    //assert(!pkt->isWrite());
+    //assert(mem_pkt->isRead());
```
Comment (powerjg) [inline src/mem/mem_ctrl.cc:266]: Re-add this assert

File: src/mem/mem_ctrl.cc [hunk 30]
```diff
@@ -187,24 +198,120 @@ MemCtrl::writeQueueFull(unsigned int neededEntries) const
...
+MemCtrl::dramFillQueueFull(unsigned int neededEntries) const
+{
+    auto size = (dramFillQueueSize + neededEntries);
+    // random size to compare with for now
+    //return  size_new > 64;
+    return  size > 64;
+}
+
+void
+MemCtrl::addToNVMReadQueue(MemPacket* mem_pkt)
```
Comment (powerjg) [inline src/mem/mem_ctrl.cc:253]:
```suggestion
MemCtrl::addToNVMReadQueue(const MemPacket* mem_pkt)
```

File: src/mem/mem_ctrl.cc [hunk 31]
```diff
@@ -187,24 +198,120 @@ MemCtrl::writeQueueFull(unsigned int neededEntries) const
...
+    //assert(!pkt->isWrite());
+    //assert(mem_pkt->isRead());
+    assert(!nvmReadQueueFull(1));
+
+    // TODO: delete the old mem pkt object
+    // COMMENT: Should overwrite the mem_pkt?
+    // COMMENT: Currently, we are assuming that
+    // a pkt will be decomposed in only one mem_pkt
+    // which probably is not a reasonable assumption
+    mem_pkt = nvm->decodePacket(mem_pkt->pkt, mem_pkt->pkt->getAddr(),
```
Comment (powerjg) [inline src/mem/mem_ctrl.cc:274]: New variable.
```suggestion
    nvm_pkt = nvm->decodePacket(mem_pkt->pkt, mem_pkt->pkt->getAddr(),
```

File: src/mem/mem_ctrl.cc [hunk 32]
```diff
@@ -302,10 +460,90 @@ MemCtrl::addToReadQueue(PacketPtr pkt, unsigned int pkt_count, bool is_dram)
...
+
+    //MARYAM: I guess the next line must assert for WRITES, not READS!
+    //assert(mem_pkt->isWrite());
+    assert(!dramFillQueueFull(1));
+
+    // COMMENT: Should overwrite the mem_pkt?
+    // COMMENT: Currently, we are assuming that
+    // a pkt will be decomposed in only one mem_pkt
+    // which probably is not a reasonable assumption
+    mem_pkt = dram->decodePacket(mem_pkt->pkt, mem_pkt->pkt->getAddr(),
```
Comment (powerjg) [inline src/mem/mem_ctrl.cc:485]:
```suggestion
    fill_pkt = dram->decodePacket(mem_pkt->pkt, mem_pkt->pkt->getAddr(),
```

File: src/mem/mem_ctrl.cc [hunk 33]
```diff
@@ -302,10 +460,90 @@ MemCtrl::addToReadQueue(PacketPtr pkt, unsigned int pkt_count, bool is_dram)
...
         DPRINTF(MemCtrl, "Request scheduled immediately\n");
+        //COMMENT: DRAM cache tag check latency should be added to the
+        // curTick()
+        schedule(nextReqEvent, curTick());
+        //COMMENT: tags will be checked on the way back
+    }
+}
+
+void
+MemCtrl::addToDRAMFillQueue(MemPacket mem_pkt)
```
Comment (powerjg) [inline src/mem/mem_ctrl.cc:471]:
```suggestion
MemCtrl::addToDRAMFillQueue(const MemPacket *mem_pkt)
```

Check on pointer vs not pointer.

File: src/mem/mem_ctrl.cc [hunk 34]
```diff
@@ -489,13 +844,343 @@ MemCtrl::recvTimingReq(PacketPtr pkt)
...
+                        retryNVMRdReq = true;
+                    if (dramFillQueueFull(1))
+                        retryDRAMFillReq = true;
+
+                    //schedule(respondEvent,
+                    //              respQueue.top().second->readyTime + 2);
+                    // re schedule the respondEvent process
+                    // not sure what should be the delay above and if even
+                    // this method would work as there might be overlap with
+                    // other events already scheduled for resp queue
```
Comment (powerjg) [inline src/mem/mem_ctrl.cc:919]: In function?

File: src/mem/mem_ctrl.cc [hunk 35]
```diff
@@ -489,13 +844,343 @@ MemCtrl::recvTimingReq(PacketPtr pkt)
...
+                    addToNVMReadQueue(mem_pkt);
+                    addToDRAMFillQueue(mem_pkt);
+                } else {
+                    // if any of the queues are successful
+                    assert(respQueue.top().second->readyTime ==
+                                                    respQueue.top.first);
+                    if (nvmReadQueueFull(1))
+                        retryNVMRdReq = true;
+                    if (dramFillQueueFull(1))
+                        retryDRAMFillReq = true;
```
Comment (powerjg) [inline src/mem/mem_ctrl.cc:912]: Remember { }

File: src/mem/mem_ctrl.cc [hunk 36]
```diff
@@ -517,16 +1202,23 @@ MemCtrl::processRespondEvent()
         }
     } else {
         // it is not a split packet
```
Comment (powerjg) [inline src/mem/mem_ctrl.cc:1204]: Add a panic to the above code in case you have a split packet since this code doesn't handle it correctly for now

File: src/mem/mem_ctrl.cc [hunk 37]
```diff
@@ -302,10 +460,90 @@ MemCtrl::addToReadQueue(PacketPtr pkt, unsigned int pkt_count, bool is_dram)
...
+    int index = bits(mem_pkt->pkt->getAddr(),
+            ceilLog2(64)+ceilLog2(numEntries), ceilLog2(64));
+
+    tagStoreDC[index].tag = returnTag(mem_pkt->pkt->getAddr());
+
+    // make sure that the block is set to be valid and clean
+    // MARYAM: only valid bit is set here. How about dirty bit?
+    tagStoreDC[index].metadata = 0x01;
+
+    //isInWriteQueue.insert(burstAlign(addr, is_dram));
```
Comment (powerjg) [inline src/mem/mem_ctrl.cc:507]: This is to speed up snooping, I think. you could create your own `isInDramFillQueue` which works similarly.

File: src/mem/mem_ctrl.cc [hunk 38]
```diff
@@ -302,10 +460,90 @@ MemCtrl::addToReadQueue(PacketPtr pkt, unsigned int pkt_count, bool is_dram)
...
+MemCtrl::addToNVMWriteQueue(MemPacket* mem_pkt)
+{
+    assert(pkt->isWrite());
+    assert(!isNVMWriteQueueFull());
+
+    // COMMENT: Should overwrite the mem_pkt?
+    // COMMENT: Currently, we are assuming that
+    // a pkt will be decomposed in only one mem_pkt
+    // which probably is not a reasonable assumption
+    mem_pkt = nvm->decodePacket(mem_pkt->pkt, mem_pkt->pkt->getAddr(),
```
Comment (powerjg) [inline src/mem/mem_ctrl.cc:527]:
```suggestion
    nvm_pkt = nvm->decodePacket(mem_pkt->pkt, mem_pkt->pkt->getAddr(),
```

File: src/mem/mem_ctrl.cc [hunk 39]
```diff
@@ -302,10 +460,90 @@ MemCtrl::addToReadQueue(PacketPtr pkt, unsigned int pkt_count, bool is_dram)
...
+    if (!nextReqEvent.scheduled()) {
+        // We might need to add tag check latency
+        // because the NVM response might check tags
+        // to see if dram miss was clean or dirty
         schedule(nextReqEvent, curTick());
     }
 }
 
+void
+MemCtrl::addToNVMWriteQueue(MemPacket* mem_pkt)
```
Comment (powerjg) [inline src/mem/mem_ctrl.cc:518]:
```suggestion
MemCtrl::addToNVMWriteQueue(const MemPacket* mem_pkt)
```

File: src/mem/mem_ctrl.cc [hunk 40]
```diff
@@ -302,10 +460,90 @@ MemCtrl::addToReadQueue(PacketPtr pkt, unsigned int pkt_count, bool is_dram)
...
+    nvmWriteQueue.push_back(mem_pkt);
+
+    nvmWriteQueueSize++;
+
+    if (!nextReqEvent.scheduled()) {
+    // COMEMNT: scheduling a request if it has not been
+    // previously scheduled.
+    // If something else has scheduled nextReqEvent, how would
+    // we add tagCheckLatency. And does it even matter to add that
+    schedule(nextReqEvent, curTick() + tagCheckLatency);
```
Comment (powerjg) [inline src/mem/mem_ctrl.cc:541]: whitespace
```suggestion
        schedule(nextReqEvent, curTick() + tagCheckLatency);
```

Either add a parameter to this function with latency or explain why you add tagCheckLatency here and not in the other fill/add to queue functions.

File: src/mem/mem_ctrl.cc [hunk 41]
```diff
@@ -489,13 +844,343 @@ MemCtrl::recvTimingReq(PacketPtr pkt)
...
+                    //              respQueue.top().second->readyTime + 2);
+                    // re schedule the respondEvent process
+                    // not sure what should be the delay above and if even
+                    // this method would work as there might be overlap with
+                    // other events already scheduled for resp queue
+                }
+            } else { // write packet
+                if (!writeAllocatePolicy) { // false = no allocate on writes
+                    if (!nvmWriteQueueFull(1) ) {
+                        addToNVMWriteQueue(mem_pkt);
```
Comment (powerjg) [inline src/mem/mem_ctrl.cc:924]: I'd add a latency parameter to these functions.

File: src/mem/mem_ctrl.cc [hunk 42]
```diff
@@ -335,18 +573,31 @@ MemCtrl::addToWriteQueue(PacketPtr pkt, unsigned int pkt_count, bool is_dram)
...
+            // the mem_pkt returned has interface specific things (different)
+            // for different devices
+            // If this is not serviced by DRAM let's say, then you will need
+            // a way to re-write the packet in the write queue.
+
+            // COMMENT: Check if this packet is in DRAM cache through tags
             if (is_dram) {
                 mem_pkt = dram->decodePacket(pkt, addr, size, false, true);
+                mem_pkt->read_before_write = true;
                 dram->setupRank(mem_pkt->rank, false);
```
Comment (powerjg) [inline src/mem/mem_ctrl.cc:588]: Should this be true?

File: src/mem/mem_ctrl.cc [hunk 43]
```diff
@@ -335,18 +573,31 @@ MemCtrl::addToWriteQueue(PacketPtr pkt, unsigned int pkt_count, bool is_dram)
...
             } else {
                 mem_pkt = nvm->decodePacket(pkt, addr, size, false, false);
```
Comment (powerjg) [inline src/mem/mem_ctrl.cc:590]: Add panic?

File: src/mem/mem_ctrl.cc [hunk 44]
```diff
@@ -412,6 +674,14 @@ MemCtrl::printQs() const
 #endif // TRACING_ON
 }
 
+inline Addr
+MemCtrl::returnTag(Addr request_addr)
+{
+    int index_bits = ceilLog2(numEntries);
+    int block_bits = ceilLog2(64);
+    return request_addr >> (index_bits+block_bits);
```
Comment (powerjg) [inline src/mem/mem_ctrl.cc:682]: Use bits()

File: src/mem/mem_ctrl.cc [hunk 45]
```diff
@@ -481,6 +834,8 @@ MemCtrl::recvTimingReq(PacketPtr pkt)
             return false;
         } else {
             addToReadQueue(pkt, pkt_count, is_dram);
+
+            // COMMENT: revisit stats
```
Comment (powerjg) [inline src/mem/mem_ctrl.cc:838]: Do it!

File: src/mem/mem_ctrl.cc [hunk 46]
```diff
@@ -939,11 +1681,12 @@ MemCtrl::processNextReqEvent()
         // track if we should switch or not
         bool switch_to_writes = false;
 
-        if (totalReadQueueSize == 0) {
+        if (totalReadQueueSize == 0 && nvmReadQueueSize == 0) {
```
Comment (powerjg) [inline src/mem/mem_ctrl.cc:1684]: Why this change?

File: src/mem/mem_ctrl.cc [hunk 47]
```diff
@@ -930,6 +1670,8 @@ MemCtrl::processNextReqEvent()
         // if all ranks are refreshing wait for them to finish
         // and stall this state machine without taking any further
         // action, and do not schedule a new nextReqEvent
+
+        // COMMENT: Then who schedules and how the nextReqEvent is scheduled
```
Comment (powerjg) [inline src/mem/mem_ctrl.cc:1674]: Looks like the below may be triggered when you don't expect it.

File: src/mem/mem_ctrl.cc (hunk 24 above)
Comment (mbabaie) [reply src/mem/mem_ctrl.cc:207]: Should these vars (e.g. maxNvmWriteQueueSize) be 'Param' or regular const vars?

Comment (powerjg) [reply src/mem/mem_ctrl.cc:207]: I'd go ahead and make them params... You'll probably want this at some point anyway.

File: src/mem/mem_ctrl.cc (hunk 23 above)
Comment (ayaz91) [reply src/mem/mem_ctrl.cc]: done.

File: src/mem/mem_ctrl.cc (hunk 42 above)
Comment (ayaz91) [reply src/mem/mem_ctrl.cc:465]: I think this should be true for all writes

File: src/mem/mem_ctrl.cc (hunk 44 above)
Comment (ayaz91) [reply src/mem/mem_ctrl.cc]: done.

File: src/mem/mem_ctrl.cc (hunk 46 above)
Comment (ayaz91) [reply src/mem/mem_ctrl.cc:1586]: This is to make sure that we do not switch to writes only because ReadQueue is empty (nvmReadQeueue can still have some requests)
